src/main/scala/com/microsoft/partnercatalyst/fortis/spark/analyzer/timeseries/ComputedAnomalyKmeansModelInputDStream.scala

```diff
@@ -0,0 +1,149 @@
+    }
+  }
+
+  override def compute(validTime: Time): Option[RDD[MLModel]] = {
+    val spark = session
+    import spark.implicits._
+
+    val trustedSourcesTable = session.sparkContext.cassandraTable[TrustedSource]("fortis", "trustedsources").cache()
+    val supportedSourcesTuples = trustedSourcesTable.map(s=>(s.pipelinekey, s.externalsourceid)).collect().toSet
+
+    val periodTypeName = periodType.periodTypeName
+    val periods = this.periodType.retrospectivePeriods(new Date().getTime, this)
+    val topicCounts = session.sparkContext.cassandraTable[TopicCount]("fortis", "topiccounts")
+        .where(
+          "periodtype = ? and period in ?",
+          periodTypeName, periods
+        )
+        .filter(tile=>{
+          tile.conjunctiontopics._2.isEmpty &&
+          tile.conjunctiontopics._3.isEmpty &&
+            (tile.externalsourceid == "all" || supportedSourcesTuples.contains((tile.pipelinekey, tile.externalsourceid)))
+        })
+        .cache()
+
+    val uniqueSourcesCount = topicCounts.toDF().select("pipelinekey", "externalsourceid").distinct().count()
+    val uniqueTopicsCount = topicCounts.toDF().select("conjunctiontopics").distinct().count()
+
+    val pt = this.periodType
+    val k = ComputedAnomalyKmeansInputRecord.getK(pt, uniqueSourcesCount.toInt, uniqueTopicsCount.toInt)
+    val kmeans = new KMeans()
+    kmeans.setK(k)
+    kmeans.setMaxIter(20)
+    kmeans.setFeaturesCol("features")
+    kmeans.setPredictionCol("prediction")
+
+    val inputs = topicCounts.map(tile=>ComputedAnomalyKmeansInputRecord(
+      tile.period,
+      periodTypeName,
+      tile.pipelinekey,
+      tile.externalsourceid,
+      features = ComputedAnomalyKmeansInputRecord.features(pt, tile)
+    )).toDF()
+    val model = kmeans.fit(inputs)
+    val WSSSE = model.computeCost(inputs)
+
+    Some(this.ssc.sparkContext.parallelize(Seq(
+      MLModel(
+        s"kmeans.${periodTypeName}.day_of_month",
+        Map(
+          ("type", model.getClass.getSimpleName),
+          ("cost", WSSSE.toString)
+        ),
+        MLModel.dataFrom(model)
+      )
+    )))
+  }
+
+  override def retrospectivePeriodCountPerType(): Map[String, Int] = Map(
+    (PeriodType.Minute.periodTypeName, 4*60), /* 4 hours */
+    (PeriodType.Hour.periodTypeName, 7*24), /* 7 days */
+    (PeriodType.Day.periodTypeName, 365), /* 1 year */
+    (PeriodType.Week.periodTypeName, 2*52), /* 2 years */
+    (PeriodType.Month.periodTypeName, 3*12), /* 3 years */
+    (PeriodType.Year.periodTypeName, 7) /* 7 years */
+  )
+
+}
+
+private object ComputedAnomalyKmeansInputRecord extends Serializable {
+
+  def getK(periodType: PeriodType,
+           uniqueSourcesCount: Int,
+           uniqueTopicsCount: Int): Int = {
+    var k = 31 /* max days in a month */
+    k *= uniqueSourcesCount
+    k *= uniqueTopicsCount
+    k
+  }
+
+  def features(periodType: PeriodType,
+               topicCount: TopicCount): Vector = {
+    val calendar = new GregorianCalendar()
```

Comment:
The old Java Calendar API is [pretty broken](https://stackoverflow.com/q/1571265/3817588). I'd recommend against using it and use the [Java8 Joda-time API](https://docs.oracle.com/javase/8/docs/api/java/time/ZonedDateTime.html#getDayOfMonth--) instead which is safer and less verbose:

```java
ZonedDateTime.ofInstant(Instant.ofEpochMilli(topicCount.periodstartdate), ZoneId.of("UTC")).getDayOfMonth()
```